### PR TITLE
New -f Option, New -s Option, Replacing run With runAll

### PR DESCRIPTION
**Pull request by lyndonhenry**
### New `-s/--stratify` option

- enables compilation to multiple subprograms, each a stratum of the original input datalog program
- this means subprograms themselves are individual semipositive datalog programs, and can be executed in isolation if given the outputs of all relevant strata preceding them in some valid order
- 'some valid order' here means the partial order induced by the SCC graph (i.e. each stratum is an SCC)
- the 'SCC' graph is written to a file in either 'dot' or 'csv' format
- see the file-by-file items on `src/RamExecutor.h` and `src/AstAnalysis.h` for more information

### New `-f/--fault-tolerance` option

- enables fault tolerant behavior for Souffle programs
- if a program terminates mid-way through execution, it can be re-run without loss of computation
- it does this by treating every relation possible as both an input and output relation
- relations involving records cannot be input relations, and so this behavior does not apply for them
- these 'intermediate' relations are written to the directory passed with the `-D` option
- they are given a `.facts` extension so as not to be confused with real outputs
- 'file not found' errors stay the same, as none will be thrown if an intermediate relation's file is not found when being used as an input relation, however still will be for true input relations

### New `runAll` function to replace `run` in generated code

- previously, all loads occurred in the `loadAll` function, computation in the `run` function, and stores/printsizes were done in the `printAll` function
- the main function executed in the order of 1.`loadAll` 2.`run` 3.`printAll`, meaning all IO occurred exclusively at the beginning and end of computation
- in this pull request, loads, stores, and printsize operations occur in the order they are given in the RAM program
- the RAM program can be found in the debug report for a program (`-r` option)
- this means that loads, stores, and printsize operations occur alternatively with computation
- as a consequence, output relations can be dropped from memory earlier, while input relations loaded earlier
- this minimizes memory consumption under the constraint of the topological order chosen for the SCCs

### Other

- much renaming
- better use of utility methods
- easier debugging of generated code
- misc. code quality improvements

## File By File

- `.gitignore`: 
Moved content of `tests/.gitignore` to `.gitignore`, as it seems a simpler way of doing things to have all ignores consolidated in a single file.
- `.souffle-sh`: 
See list item for `souffle.sh`.
- `.travis.yml`: 
Added test cases for new `-f` and `-s` options, and changed all existing options to use multiple threads. The latter of these is not so much for the potential speedup, but because it happens more often that a bug will appear in multithreaded but not single threaded execution than vise-versa. Also added one `TODO` to note that running the Evaluation tests with the `-s` option is not possible with the time limit available on Travis. See issue #468 for more information.
- `souffle.sh`: 
Renamed `.souffle.sh`, also has been reduced from 10 commands to 3, while retaining the same functionality. 
- `src/AstAnalysis.h`: 
Added `print` method to `AstAnalysis` base class, and implemented for relevant subclasses to replace their `outputMyClass`-style methods. For example, `topsortSCCGraph->outputTopologicallySortedSCCGraph()` becomes the much more readable `topsortSCCGraph->print()`.
- `src/AstParserUtils.cpp`: 
Formatting fixes.
- `src/AstRelation.h`: 
Added method to check if an `AstRelation` has an argument in its head that is a record type. This is used in a half-implemented solution for allowing subprogram compilation for programs involving record types. See the entry for `src/AstSemanticChecker.cpp` for more information.
- `src/AstSemanticChecker.cpp`: 
Method name changes as per the refactoring of `src/PrecedenceGraph.h`, `src/PrecedenceGraph.cpp` and `src/GraphUtils.h`. Automatic disabling of `-s` option in the presence of record types. Added two `TODO`'s to help in fixing subprogram compilation involving record types. See issue #467 for more information.
- `src/CompiledRamIndexUtils.h`:
Formatting fixes.
- `src/ComponentModel.cpp`:
Formatting fixes.
- `src/DebugReport.cpp`: 
Changed subclasses of `AstAnalysis` to now use the `print` method, as per the changes noted in `src/AstRelation.h`.
- `src/Global.h`:
Added `unset` method to enable option disabling as per the changes in `src/AstSemanticChecker.cpp`.
- `src/GraphUtils.h`: 
Change of method names for consistency with `src/PrecedenceGraph.h`, and also for clarity. For example, `graph->predecessors(x)` is more obvious and easy to remember than `graph->getReverseEdges(x)`. 
- `src/IOSystem.h`: 
Renaming as per the changes in `src/ReadStreamSQLite.h`.
- `src/MagicSet.cpp`: 
Renaming as per the changes in `src/AstAnalysis.h`.
- `src/MagicSet.h`: 
Renaming as per the changes in `src/AstAnalysis.h`.
- `src/ParserDriver.cpp`:
Formatting fixes.
- `src/PrecedenceGraph.cpp`: 
Renaming as per `src/PrecedenceGraph.h` and `GraphUtils.h`.  Renaming and implementation of `print` methods as per `src/AstAnalysis.h`. Implementation of parameterized `print` method for `SCCGraph` class. When compiled with the new `-s` option, Souffle will write the SCC graph to the file given by the path passed as the argument to the option, in the format of the extension of the file of that path. Currently supported formats are given in the help text of `src/main.cpp`. If the argument passed is a `-`, then each subprogram executable is executed in the order of the indices as described in the list item for `src/RamExecutor.h`. Added one `TODO` to help in solving problems with subprogram compilation involving records. See issue #467 for more information.
- `src/PrecedenceGraph.h`:  
Various renaming of functions and variables for consistency between all classes defined in the file in question, as well as `src/GraphUtils.h`. Removal of several unused functions. See the changes in `src/PrecedenceGraph.cpp` for more information.
- `src/RamExecutor.cpp`: 
Added `/* BEGIN ... */` and `/* END ... */` comments for visited RAM types to the generated code; note that these only show up with the `--verbose` or `--debug-report` options enabled. Identified main source of bug for when subprogram compilation involving records is enabled by introducing artificial dependencies to the precedence graph, so added a `TODO`. See issue #467 for more information. Implemented `visitLoad`, `visitStore`, and `visitPrintSize` methods for code generation, which are called during the `runAll` method only (i.e. not if `run` is called). Added `runAll` method in generated code, which is now called by `main` instead of `run`.  `runAll` performs loads and stores alternatively with computation, in an execution order reflecting the RAM program printed in the debug report.  `RamDrop`'s now also result in `purge` calls for non-temporary relations, freeing memory earlier. Together with the new `runAll` method, this effectively minimizes memory usage under the constraint of the chosen topological order of the SCCs.
- `src/RamExecutor.h`:  
Addition of index parameter to `generateCode`, `compileToLibrary`, and `compileToBinary` functions. This is used by the new `-s` option to number the produced subprograms, where the passed index is appended to the filename of the output file, preceded by an underscore. If this index is -1 (as it is by default when no `-s` option is given), no index is appended and behavior does not change. The indices themselves give a topological order over the execution graph of the produced executables, meaning that if these executables are run in order of their appended indices, the execution will be correct (assuming it is under the same configuration when the `-s` option is not given). If `-s` is given with the `-` argument, then such an execution is exactly what takes place. See the changes in `src/RamExecutor.cpp` for more information.
- `src/RamInterface.h`: 
Added `runAll` method as per the changes in `src/SouffleInterface.h`. Removed pointless error messages.
- `src/RamStatement.h`: 
Added support for printing with the `<<` operator, printed RAM program now lets user know the file path of loads and stores, merge statement now prints relations in destination-source order to match the class constructor (previously was source-destination order) with clearer wording of what gets merged in to what.
- `src/RamTranslator.cpp`: 
Added ability to specify path and extension when retreiving a RamRelationIdentifier, to allow for the new scenario introduced by the `-f` and `-s` options where a `.facts` file is written to the output directory. Significant changes to `RamTranslator::translateProgram` method, to account for the different scenarios of the new `-s` and `-f` options. Added two `TODO`'s to help with the future possibility of storing new relations immediately as they are discovered. See issue #466 for more information.
- `src/RamValue.h`:
Formatting fixes.
- `src/ReadStreamCSV.h`: 
Ensure that attempted reads of files for intermediate relations fail silently if a file is not found, to allow for restarts upon failure if the new `-s` and `-f` options are used simultaneously. Note that the effect of this is that programs compiled with both `-s` and `-f` will give the same 'file not found' errors on the same inputs as programs compiled or interpreted under any other configuration of options, and so the testsuite can be used without modification.
- `src/ReadStreamSQLite.h`: 
Naming consistency, `WriteSQLiteFactory` is to `WriteFileCSVFactory` as `ReadSQLiteFactory` is to `ReadFileCSVFactory` (`ReadSQLiteFactory` was previously called `ReadStreamSQLiteFactory`).
- `src/SouffleInterface.h`: 
Added `runAll` method. This method takes an input and output directory as arguments; it perfoms loads for input relations just before they are used, and performs stores for output relations just before they are expired. All behaviour for existing methods has been preserved. Note that `runAll` is now called instead of `run` in the generated code. See the list item for `src/RamExecutor.cpp` for more information.
- `src/Util.h`: 
Addition of a few utility methods for files and identifiers moved over from other parts of the code where they didn't really belong.
- `src/main.cpp`: 
Addition of `-s/--stratify` and `-f/--fault-tolerance` options to list of main options. Changes to allow for `-s` option, see list item for `src/RamTranslator.cpp` for more information.
- `src/test/graph_utils_test.cpp`: 
Refactoring for the changes in `src/GraphUtils.h`.
- `src/test/test.h`:
Formatting fixes.
- `tests/.gitignore`: 
Removed, see list item `.gitignore`.
- `tests/evaluation.at`
Added for `cpp_keywords` test, see list item for `tests/evaluation/cpp_keywords/cpp_keywords.dl`.
- `tests/evaluation/cpp_keywords/cpp_keywords.dl`
This is the previous version of `tests/syntactic/cpp_keywords/cpp_keywords.dl`, see its list item for more information.
- `tests/evaluation/cpp_keywords/cpp_keywords.err`
Added for `cpp_keywords` test, see list item for `tests/evaluation/cpp_keywords/cpp_keywords.dl`.
- `tests/evaluation/cpp_keywords/cpp_keywords.out`
Added for `cpp_keywords` test, see list item for `tests/evaluation/cpp_keywords/cpp_keywords.dl`.
- `tests/evaluation/cpp_keywords/worked?.csv`
Added for `cpp_keywords` test, see list item for `tests/evaluation/cpp_keywords/cpp_keywords.dl`.
- `tests/example/tc/facts/base.facts`: 
Added some facts as per the changes in `tests/example/tc/tc.dl`.
- `tests/example/tc/tc.dl`: 
Making one of the relations an input relation lets this test serve as a nice minimal demo of the core Souffle syntax for pretty much the most ubiquitous datalog example there is.
- `tests/syntactic/cpp_keywords/cpp_keywords.dl`:
Moved existing version of this test to  `tests/evaluation/cpp_keywords/cpp_keywords.dl`, and re-wrote new version to use record types instead of numbers. The same checks on syntax are preserved, however the execution time for the new `-s` option is significantly reduced. This allows all tests in the syntactic category to pass travis, while before this test caused a timeout.
- `tests/testsuite.at`: 
Added new tests and modified old ones in the same manner as for `.travis.yml`.

## Edits
- In addition, profiling tests have been removed for evaluation tests in `.travis.yml` due to problems with timeouts; they have, however, been left in for non-evaluation tests. This is appropriate as the profiling *output*  itself is not checked by the tests, only the correctness of the execution of these tests (in terms of their outputs) with the profiling option enabled.

### CODE REVIEW

File: src/RamExecutor.cpp [hunk 1]
```diff
@@ -2246,12 +2358,12 @@ std::string RamCompiler::generateCode(
     if (Global::config().has("profile")) {
         os << "(std::string pf=\"profile.log\") : profiling_fname(pf)";
         if (initCons.size() > 0) {
-            os << ",\n";
+            os << ",\n" << initCons;
         }
     } else {
-        os << "() : \n";
+        os << "()";
+        if (initCons.size() > 0) os << " : " << initCons;
```
Comment (mmcgr) [inline src/RamExecutor.cpp]: Use brackets around the if body.

File: src/RamExecutor.cpp [hunk 2]
```diff
@@ -2511,11 +2625,11 @@ std::string RamCompiler::generateCode(
 }
 
 std::string RamCompiler::compileToLibrary(
-        const SymbolTable& symTable, const RamProgram& prog, const std::string& filename) const {
-    std::string source = generateCode(symTable, prog, filename + ".cpp");
+        const SymbolTable& symTable, const RamProgram& prog, const std::string& filename, const int index) const {
+    std::string source = generateCode(symTable, prog, filename, index);
```
Comment (mmcgr) [inline src/RamExecutor.cpp]: Could you rename `source` to, say, `sourceFilename`? It's particularly unclear when it's assigned to be a method named `generateCode`.

File: src/RamExecutor.cpp [hunk 3]
```diff
@@ -2567,12 +2680,14 @@ std::string RamCompiler::compileToBinary(const SymbolTable& symTable, const RamP
     }
 
     // done
-    return binary;
+    return source;
```
Comment (mmcgr) [inline src/RamExecutor.cpp]: Are you returning the filename of the binary produced or of the source code file? (The naming in this method was distinctly unclear well before you touched it.)

File: src/RamTranslator.cpp [hunk 4]
```diff
@@ -776,7 +787,13 @@ std::unique_ptr<RamStatement> RamTranslator::translateClause(const AstClause& cl
 static void appendStmt(std::unique_ptr<RamStatement>& stmtList, std::unique_ptr<RamStatement> stmt) {
     if (stmt) {
         if (stmtList) {
-            stmtList = std::unique_ptr<RamStatement>(new RamSequence(std::move(stmtList), std::move(stmt)));
+            RamSequence* stmtSeq;
+            if ((stmtSeq = dynamic_cast<RamSequence*>(stmtList.get()))) {
+                stmtSeq->add(std::move(stmt));
+            } else {
+                stmtList =
+                        std::unique_ptr<RamStatement>(new RamSequence(std::move(stmtList), std::move(stmt)));
```
Comment (mmcgr) [inline src/RamTranslator.cpp]: make_unique would be simpler

File: src/RamTranslator.cpp [hunk 5]
```diff
@@ -1059,15 +1086,87 @@ std::unique_ptr<RamStatement> RamTranslator::translateRecursiveRelation(
...
+        if (Global::config().has("fault-tolerance")) {
+            for (const AstRelation* rel : scc) {
+                if (!rel->hasRecordInHead()) {
+                    appendStmt(res, std::unique_ptr<RamStatement>(
+                                            new RamStore(getRamRelationIdentifier(rel, &typeEnv))));
+                }
+            }
+        }
+    }
+    if (postamble) appendStmt(res, std::move(postamble));
```
Comment (mmcgr) [inline src/RamTranslator.cpp]: Use brackets around if bodies.

File: src/RamTranslator.cpp [hunk 6]
```diff
@@ -1059,15 +1086,87 @@ std::unique_ptr<RamStatement> RamTranslator::translateRecursiveRelation(
...
+            (!isComputed && !isRecursive && !loadInputOnly && rel->isInput())) {
+        std::unique_ptr<AstIODirective> directive = std::unique_ptr<AstIODirective>(new AstIODirective());
+        directive->setAsInput();
+        directive->addKVP("intermediate", "true");
+        mrel->addIODirectives(std::move(directive));
+    }
+    RamRelationIdentifier rrel = getRamRelationIdentifier(
+            mrel, &typeEnv, getRelationName(mrel->getName()), mrel->getArity(), false, dir, ext);
+    // create and load the relation at the start
+    appendStmt(current, std::unique_ptr<RamStatement>(new RamCreate(rrel)));
```
Comment (mmcgr) [inline src/RamTranslator.cpp]: make_unique<RamCreate>(rrel) should work, I think, and be simpler. Likewise the unique_ptrs below this.

File: src/RamTranslator.cpp [hunk 7]
```diff
@@ -1125,84 +1224,96 @@ std::unique_ptr<RamProgram> RamTranslator::translateProgram(const AstTranslation
...
+        /* during stratification, create and load all predecessor relations in another scc */
+        if (Global::config().has("stratify"))
+            // for each inbound relation (i.e. a predecessor in another scc)...
+            for (const AstRelation* rel : sccGraph.getInbound(sortedSCCGraph.order().at(index)))
+                createAndLoad(current, rel, typeEnv, false, false, false);
+
+        /* during fault-tolerant stratification, attempt to recover all relations computed in this step */
+        if (Global::config().has("fault-tolerance")) {
+            for (const AstRelation* rel : step.computed())
+                createAndLoad(current, rel, typeEnv, true, sccGraph.isRecursive(rel), rel->hasRecordInHead());
```
Comment (mmcgr) [inline src/RamTranslator.cpp]: Add brackets

File: src/RamTranslator.cpp [hunk 8]
```diff
@@ -1125,84 +1224,96 @@ std::unique_ptr<RamProgram> RamTranslator::translateProgram(const AstTranslation
...
+        } else {
+            for (const AstRelation* rel : step.computed())
+                createAndLoad(current, rel, typeEnv, true, sccGraph.isRecursive(rel), true);
```
Comment (mmcgr) [inline src/RamTranslator.cpp]: Add brackets

**Comment by lyndonhenry** [comment]
Yeah, I know, I'm sorry, it really should have been done as three separate pull requests. I'd been working on multiple things at once, and was trying to get rid of all the `TODO`'s before I did a PR. When I realised the amount of changes I'd made I decided that I had to stop as there was too much in there already. In future I'll make sure to submit much smaller changes. 
There's a few minor bugs introduced by the merge with #459 that I'll have to fix as well (it was accepted after this one was submitted), but otherwise all the recommendations I have no qualms with and so I'll do them and the bugfixes tomorrow. 
Apolagies again for the confusing and disorganised behemoth that is this PR.

**Pull request merged**